### ISSUE TITLE
IEnumerable is eagerly iterated

### ISSUE THREAD

**Issue by jods4** [scriban/scriban#428]
Writing a `for` loop over `IEnumerable` objects works but it seems like it enumerates everything (supposedly in an array buffer?) before executing any loop body and performing rendering.

As a user, this is surprising as the similar `foreach` goes through the enumerable lazily.

It has some undesirable consequences:

When the template code influences the enumeration itself. 
This is my use-case and how I noticed this:
I have a queue of items that needs to be rendered, and rendering one item can enqueue a new one.
Anything enqueued inside the loop isn't processed.

Processing infinite sequences is impossible.
You may be tempted to do a `TakeWhile` pattern but it won't work:
```
for item in infinite
  if item.stop_here; break; end
end
```

**Comment by xoofx** [comment]
Agreed, it was introduced quite some time ago because Liquid was requiring the length to be accessible in a variable, so I made quite some change back in the days (the previous implementation was much simpler actually)

https://github.com/scriban/scriban/blob/3ba3cc3ec8baa65a3a5983c1e610ad5cfc662b5f/src/Scriban/Syntax/Statements/ScriptForStatement.cs#L114-L116


So rethinking about it, it should be possible to make the access of the length deferred (so that in that case you pay the price of a potential forced iteration on all elements) and use the enumeration otherwise for the loop.

**Comment by xoofx** [comment]
Fixed by commit b39f7a1, will be available in next version

**Comment by jods4** [comment]
Awesome, thanks! 
It will simplify my templates a lot, right now I fell back to a `while true` loop and a `dequeue` method.

**Comment by jods4** [comment]
@xoofx The fix doesn't seem to work in my template.
I use Scriban 5.4.4.0 and here's a small repro I've tried:

In my C# code, I introduce this built-on object that I expose as `queue`:
```csharp
class QueueBuiltin : ScriptObject
{
  static Queue<string> queue = new();

  public static void Add(string x) => queue.Enqueue(x);

  public static IEnumerable<string> Flush()
  {
    while (queue.TryDequeue(out var x))
      yield return x;
  }
}
```

I would expect that anything enqueued during a `Flush()` enumeration would be picked up if the `IEnumerable` is lazily iterated.

Here's a very short Scriban template to test it out:
```scriban
{{
  queue.add "a"
  for x in queue.flush
    x
    if x == "a"; queue.add "b"; end
  end
  "1st loop ended, queue contents:"
  for x in queue.flush
    x
  end
}}
```

But the outcome is not what I'd expect:
```
a
1st loop ended, queue contents:
b
```

**Comment by xoofx** [comment]
> @xoofx The fix doesn't seem to work in my template.

That's because you are doing something enough different that it is not working well in that case. 🙂 

The loop in Scriban has to set a variable `for.last`, so it will fetch the next element before entering the loop in order to set correct `last`:

https://github.com/scriban/scriban/blob/34396529f637455bc059d609d18630d1d1b57547/src/Scriban/Syntax/Statements/ScriptForStatement.cs#L180

If you start with this at the beginning, it will work:

```
  queue.add "a"
  queue.add "c"
```

In the case above, because `a` is not last, it will give a chance for the loop to execute before "c" and it will push "b".

**Comment by jods4** [comment]
So if I understood correctly:
it is iterating lazily but it actually always reads one element ahead of the current one?

It's understandable because there's no other way to know if one element is last or not,
but at the same time it is very surprising and reduces the usefulness of lazy iteration.

It's a tricky problem...
Do you think that you could you make `last` a lazy property? 
So if the template doesn't read `last`, it works as expected.
When the template reads `last`, well it has to assume this behavior because there's just no other way.

**Comment by xoofx** [comment]
> Do you think that you could you make last a lazy property?

Yep, it should be possible to have this tradeoff.

**Comment by xoofx** [comment]
It should be fixed by 5d192a7 and available in 5.4.5+

**Comment by jods4** [comment]
You're awsome thanks! ❤️ 
Will try it out on Monday.
